include/cuco/detail/open_addressing/kernels.cuh

```diff
@@ -359,6 +359,49 @@ CUCO_KERNEL __launch_bounds__(BlockSize) void find(InputIt first,
   }
 }
 
+// TODO docs
+template <bool IsOuter,
+          int32_t BlockSize,
+          class InputProbeIt,
+          class OutputProbeIt,
+          class OutputMatchIt,
+          class AtomicCounter,
+          class Ref>
+CUCO_KERNEL __launch_bounds__(BlockSize) void retrieve(InputProbeIt input_probe,
+                                                       cuco::detail::index_type n,
+                                                       OutputProbeIt output_probe,
+                                                       OutputMatchIt output_match,
+                                                       AtomicCounter* atomic_counter,
+                                                       Ref ref)
+{
+  namespace cg = cooperative_groups;
+
+  auto const block           = cg::this_thread_block();
+  auto const blocks_in_grid  = gridDim.x;
+  auto const elems_per_block = cuco::detail::int_div_ceil(n, blocks_in_grid);
+
+  auto const block_begin_offset = block.group_index().x * elems_per_block;
+  auto const block_end_offset   = min(n, block_begin_offset + elems_per_block);
```

Comment:
This is based on the assumption that one thread will be used per input. Therefore, the active block is doing more work than expected while the idle block is doing nothing but wait

Reply:
Not sure I understand. The idea here is that we launch a grid with a fixed number of blocks. We partition the input range evenly among all blocks and pass the sub-ranges to each block. The actual assignment from input elements to CGs happens inside the `retrieve` device function itself by using a block stride loop.

Reply:
I'm 90% sure I fixed it